# Введение в Data Science и машинное обучение
# _Задачи_

## 1. Введение

### 1.4. Pandas, Dataframes, нет, панды тут ни при чем

А теперь используем эти методы на знаменитом [титаническом датасэте](https://stepik.org/media/attachments/course/4852/titanic.csv)! 

Загрузите датасэт, посмотрите на датафрэйм и ответьте на вопросы

Число колонок в представленном датафрэйме `_`, а число строк `_`. Тип float имеют `_` колонки, int - `_`, и object `_`

In [1]:
import pandas as pd

In [2]:
titanic = pd.read_csv('data\\titanic.csv')

In [3]:
titanic.shape

(891, 12)

In [4]:
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [5]:
titanic.dtypes.value_counts()

object     5
int64      5
float64    2
dtype: int64

Число колонок в представленном датафрэйме `12`, а число строк `891`. Тип float имеют `2` колонки, int - `5`, и object `5`

### 1.5. Фильтрация данных

`?`

У какой доли студентов из [датасэта](https://stepik.org/media/attachments/course/4852/StudentsPerformance.csv) в колонке `lunch` указано `free/reduced`?

In [6]:
students_performance = pd.read_csv('data\\StudentsPerformance.csv')

In [7]:
students_performance.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [8]:
students_performance.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [9]:
len(students_performance[students_performance['lunch'] == 'free/reduced']) \
    / len(students_performance)

0.355

`?`

Как различается среднее и дисперсия оценок по предметам у групп студентов со стандартным или урезанным ланчем?

In [10]:
students_performance[students_performance['lunch'] == 'free/reduced'].describe()

math score  reading score  writing score
count  355.000000     355.000000     355.000000
mean    58.921127      64.653521      63.022535
std     15.159956      14.895339      15.433823
min      0.000000      17.000000      10.000000
25%     49.000000      56.000000      53.000000
50%     60.000000      65.000000      64.000000
75%     69.000000      75.000000      74.000000
max    100.000000     100.000000     100.000000

In [11]:
students_performance[students_performance['lunch'] != 'free/reduced'].describe()

math score  reading score  writing score
count  645.000000     645.000000     645.000000
mean    70.034109      71.654264      70.823256
std     13.653501      13.830602      14.339487
min     19.000000      26.000000      22.000000
25%     61.000000      63.000000      62.000000
50%     69.000000      72.000000      72.000000
75%     80.000000      82.000000      81.000000
max    100.000000     100.000000     100.000000

In [12]:
students_performance[students_performance['lunch'] == 'free/reduced'].var()

math score       229.824270
reading score    221.871139
writing score    238.202881
dtype: float64

In [13]:
students_performance[students_performance['lunch'] != 'free/reduced'].var()

math score       186.418089
reading score    191.285560
writing score    205.620887
dtype: float64

- `студенты со стандартным ланчем имеют в среднем более хорошие оценки`

- `дисперсия оценок студентов со стандартным ланчем меньше`

### 1.6. Группировка и агрегация, ничего, скоро привыкнем

`?`

Пересчитаем число ног у героев игры Dota2! Сгруппируйте героев из [датасэта](https://stepik.org/media/attachments/course/4852/dota_hero_stats.csv) по числу их ног (колонка `legs`), и заполните их число в задании ниже.

Данные взяты [отсюда](https://api.opendota.com/api/heroes), на [этом же сайте](https://www.opendota.com/) можно найти больше разнообразных данных по Dota2.

- Число героев с "0 ног" - `__`
- Число героев с 2-мя ногами -  `__`
- Число героев с 4-мя ногами -  `__`
- Число героев с 6-ю ногами -  `__`
- Число героев с 8-ю ногами -  `__`

In [14]:
dota = pd.read_csv('data\\dota_hero_stats.csv')

In [15]:
dota.head(2)

Unnamed: 0 attack_type  id  legs localized_name                    name  \
0           0       Melee   1     2      Anti-Mage  npc_dota_hero_antimage   
1           1       Melee   2     2            Axe       npc_dota_hero_axe   

  primary_attr                                            roles  
0          agi                     ['Carry', 'Escape', 'Nuker']  
1          str  ['Initiator', 'Durable', 'Disabler', 'Jungler']

In [16]:
dota.columns

Index(['Unnamed: 0', 'attack_type', 'id', 'legs', 'localized_name', 'name',
       'primary_attr', 'roles'],
      dtype='object')

In [17]:
dota['legs'].value_counts()

2    95
0    11
4     7
6     3
8     1
Name: legs, dtype: int64

In [18]:
dota.groupby('legs').size()

legs
0    11
2    95
4     7
6     3
8     1
dtype: int64

- Число героев с "0 ног" - `11`
- Число героев с 2-мя ногами -  `95`
- Число героев с 4-мя ногами -  `7`
- Число героев с 6-ю ногами -  `3`
- Число героев с 8-ю ногами -  `1`

`?`

К нам поступили [данные](https://stepik.org/media/attachments/course/4852/accountancy.csv) из бухгалтерии о заработках Лупы и Пупы за разные задачи! 

Посмотрите у кого из них больше средний заработок в различных категориях (колонка `Type`) и заполните таблицу, указывая исполнителя с большим заработком в каждой из категорий.

In [19]:
pupaloopa = pd.read_csv('data\\accountancy.csv')
pupaloopa.head()

Unnamed: 0 Executor Type  Salary
0           0     Pupa    D      63
1           1     Pupa    A     158
2           2     Pupa    D     194
3           3     Pupa    E     109
4           4    Loopa    E     184

In [20]:
pupaloopa_mean_salary = pupaloopa.groupby(['Type', 'Executor'])[['Salary']].mean()
pupaloopa_mean_salary

Salary
Type Executor            
A    Loopa      58.000000
     Pupa      160.833333
B    Loopa     145.166667
     Pupa       77.000000
C    Loopa     154.333333
     Pupa       74.500000
D    Loopa     137.714286
     Pupa      146.500000
E    Loopa     164.000000
     Pupa      131.200000
F    Loopa     238.000000
     Pupa      136.250000

In [21]:
pupaloopa_mean_salary.unstack()

Salary            
Executor       Loopa        Pupa
Type                            
A          58.000000  160.833333
B         145.166667   77.000000
C         154.333333   74.500000
D         137.714286  146.500000
E         164.000000  131.200000
F         238.000000  136.250000

`?`

Продолжим исследование героев `Dota2`. 

Сгруппируйте по колонкам `attack_type` и `primary_attr` и выберите самый распространённый набор характеристик.

In [22]:
dota.head()

Unnamed: 0 attack_type  id  legs  localized_name  \
0           0       Melee   1     2       Anti-Mage   
1           1       Melee   2     2             Axe   
2           2      Ranged   3     4            Bane   
3           3       Melee   4     2     Bloodseeker   
4           4      Ranged   5     2  Crystal Maiden   

                           name primary_attr  \
0        npc_dota_hero_antimage          agi   
1             npc_dota_hero_axe          str   
2            npc_dota_hero_bane          int   
3     npc_dota_hero_bloodseeker          agi   
4  npc_dota_hero_crystal_maiden          int   

                                               roles  
0                       ['Carry', 'Escape', 'Nuker']  
1    ['Initiator', 'Durable', 'Disabler', 'Jungler']  
2        ['Support', 'Disabler', 'Nuker', 'Durable']  
3  ['Carry', 'Disabler', 'Jungler', 'Nuker', 'Ini...  
4        ['Support', 'Disabler', 'Nuker', 'Jungler']

In [23]:
dota.groupby(['attack_type', 'primary_attr']).count()

Unnamed: 0  id  legs  localized_name  name  roles
attack_type primary_attr                                                   
Melee       agi                   19  19    19              19    19     19
            int                    2   2     2               2     2      2
            str                   35  35    35              35    35     35
Ranged      agi                   18  18    18              18    18     18
            int                   40  40    40              40    40     40
            str                    3   3     3               3     3      3

In [24]:
dota.groupby(['attack_type', 'primary_attr'])[['id']].count()

id
attack_type primary_attr    
Melee       agi           19
            int            2
            str           35
Ranged      agi           18
            int           40
            str            3

In [25]:
dota.filter(items=['attack_type', 'primary_attr']).mode()

attack_type primary_attr
0      Ranged          int

`>_`

Аспирант Ростислав изучает метаболом водорослей и получил такую [табличку](http://stepik.org/media/attachments/course/4852/algae.csv). В ней он записал вид каждой водоросли, её род (группа, объединяющая близкие виды), группа (ещё одно объединение водорослей в крупные фракции) и концентрации анализируемых веществ.

Помогите Ростиславу найти среднюю концентрацию каждого из веществ в каждом из родов (колонка `genus`)! Для этого проведите группировку датафрэйма, сохранённого в переменной `concentrations`, и примените метод, сохранив результат в переменной `mean_concentrations`.

Модуль `pandas` уже импортирован.

In [26]:
concentrations = pd.read_csv('data\\algae.csv')
concentrations.head()

species       genus  group   sucrose    alanin   citrate  \
0    Fucus_vesiculosus       Fucus  brown  3.001472  3.711498  5.004262   
1  Saccharina_japonica  Saccharina  brown  6.731070  1.255251  5.621499   
2       Fucus_serratus       Fucus  brown  3.276870  0.346431  1.216767   
3      Fucus_distichus       Fucus  brown  6.786996  6.641303  6.423606   
4    Cladophora_fracta  Cladophora  green  3.861470  1.648450  6.940588   

    glucose  oleic_acid  
0  2.548459    6.405165  
1  6.013219    4.156700  
2  3.623225    0.304573  
3  2.272724    3.393203  
4  2.316955    2.528886

In [27]:
mean_concentrations = concentrations.groupby('genus').mean()

In [28]:
mean_concentrations

sucrose    alanin   citrate   glucose  oleic_acid
genus                                                          
Ascophyllum  6.825467  0.875429  5.253527  3.414961    2.432526
Cladophora   4.008792  3.997055  5.288311  2.800276    2.263472
Fucus        4.355112  3.566411  4.214878  2.814803    3.367647
Palmaria     0.704580  3.176440  5.573905  3.242090    2.245538
Saccharina   4.183596  3.524207  3.346710  4.980594    4.487252